Add support for TryFi Series 3+ behavior metrics and enhanced device info

adamjacobmuller opened the pull request:
# TryFi Series 3+ Home Assistant Integration Update

## Overview
This update adds support for new metrics available in the TryFi Series 3+ collar to the Home Assistant integration.

## New Sensors Added

### Behavior Sensors (Series 3+ Only)
For each behavior type, sensors are created for daily, weekly, and monthly periods:

**Barking Sensors:**
- `sensor.{pet_name}_daily_barking_count` - Number of barking episodes today
- `sensor.{pet_name}_daily_barking_duration` - Total barking duration today (seconds)
- `sensor.{pet_name}_weekly_barking_count` - Number of barking episodes this week
- `sensor.{pet_name}_weekly_barking_duration` - Total barking duration this week (seconds)
- `sensor.{pet_name}_monthly_barking_count` - Number of barking episodes this month
- `sensor.{pet_name}_monthly_barking_duration` - Total barking duration this month (seconds)

**Licking Sensors:**
- `sensor.{pet_name}_daily_licking_count` - Number of licking episodes today
- `sensor.{pet_name}_daily_licking_duration` - Total licking duration today (seconds)
- And weekly/monthly variants...

**Scratching Sensors:**
- `sensor.{pet_name}_daily_scratching_count` - Number of scratching episodes today
- `sensor.{pet_name}_daily_scratching_duration` - Total scratching duration today (seconds)
- And weekly/monthly variants...

**Eating Sensors:**
- `sensor.{pet_name}_daily_eating_count` - Number of eating sessions today
- `sensor.{pet_name}_daily_eating_duration` - Total eating duration today (seconds)
- And weekly/monthly variants...

**Drinking Sensors:**
- `sensor.{pet_name}_daily_drinking_count` - Number of drinking sessions today
- `sensor.{pet_name}_daily_drinking_duration` - Total drinking duration today (seconds)
- And weekly/monthly variants...

### Enhanced Device Information Sensors
- `sensor.{pet_name}_firmware_version` - Current firmware version of the collar
- `sensor.{pet_name}_hardware_version` - Hardware version identifier
- `sensor.{pet_name}_gps_accuracy` - GPS accuracy in meters

## Implementation Details

### Changes Made:
1. **Updated embedded pytryfi library** with Series 3+ support
2. **Added new sensor descriptions** for all behavior metrics
3. **Created PetBehaviorSensor class** to handle behavior data
4. **Extended PetGenericSensor** to support new device fields
5. **Added proper units and icons** for all new sensors

### Backward Compatibility:
- All new sensors will show `0` or `Unknown` for older collar models
- No breaking changes to existing sensors
- Integration continues to work normally with all collar versions

## Example Automations

### Alert for Excessive Barking
```yaml
automation:
  - alias: "Alert on Excessive Barking"
    trigger:
      - platform: numeric_state
        entity_id: sensor.rover_daily_barking_count
        above: 50
    action:
      - service: notify.mobile_app
        data:
          title: "Excessive Barking Alert"
          message: "Rover has barked {{ states('sensor.rover_daily_barking_count') }} times today!"
```

### Monitor Drinking Habits
```yaml
automation:
  - alias: "Low Water Consumption Alert"
    trigger:
      - platform: time
        at: "20:00:00"
    condition:
      - condition: numeric_state
        entity_id: sensor.rover_daily_drinking_count
        below: 3
    action:
      - service: notify.mobile_app
        data:
          title: "Low Water Consumption"
          message: "Rover has only drunk water {{ states('sensor.rover_daily_drinking_count') }} times today"
```

## Testing
These changes are based on anticipated API structure for Series 3+ collars. Testing with an actual Series 3+ collar is recommended to verify functionality.

adamjacobmuller commented:
## Update: Actual Implementation Based on Real API

After testing with a real Series 3+ collar, this PR has been updated to implement only the features that actually exist in the API.

### Actually Implemented Sensors

This PR now adds the following **real** sensors for Series 3+ collars:

1. **sensor.{pet_name}_battery_voltage** - Battery voltage in volts (V)
2. **sensor.{pet_name}_temperature** - Internal temperature in °C 
3. **sensor.{pet_name}_uptime** - Device uptime in seconds
4. **sensor.{pet_name}_hardware_revision** - Hardware version (e.g., "prod1")

### Removed Features (Not Available in API)

All behavior tracking sensors have been removed as these features are not exposed in the current API:
- ❌ Barking detection and sensors
- ❌ Licking detection and sensors
- ❌ Scratching detection and sensors
- ❌ Eating detection and sensors
- ❌ Drinking detection and sensors
- ❌ Firmware version sensor
- ❌ GPS accuracy sensor

### Implementation Details

- Added proper Home Assistant device classes and units for new sensors
- Battery voltage uses `UnitOfElectricPotential.VOLT` with `device_class: voltage`
- Temperature uses `UnitOfTemperature.CELSIUS` with `device_class: temperature`
- Uptime uses `UnitOfTime.SECONDS` with `state_class: total_increasing`

### Note on Temperature

The temperature value from the API appears to be a raw sensor reading. The actual conversion to Celsius may need adjustment based on the sensor specifications.

### Testing Status

✅ Tested and working with a real Series 3+ collar
✅ All new sensors properly appear in Home Assistant
✅ Backward compatible with older collars

ajacques commented:
I think this undoes some of the refactoring work I did in the pytryfi library to improve error handling and batch up graphql calls. Was that intentional? It would be good if we could get both my changes and your changes merged in safely

adamjacobmuller commented:
Just pushed a commit to fix the critical linting errors that were causing CI failures:
- Added missing json import 
- Fixed bare except statements
- Fixed TryFiError imports
- Removed duplicate property definition

The CI should have fewer errors now. There are still some star import warnings remaining, but those are less critical and would require more extensive refactoring of the import structure.

ajacques commented on custom_components/tryfi/pytryfi/__init__.py:
```diff
@@ -1,106 +1,187 @@
-                self._bases.append(b)
+
+        try:
+            # Only initialize Sentry if not in test environment
+            import os
+            if not os.environ.get("PYTEST_CURRENT_TEST"):
+                sentry_sdk.init(
+                        SENTRY_URL,
+                        release=PYTRYFI_VERSION,
+                    )
```
I don't think we should be logging to Sentry in the library for several reasons: 1. I assume we don't have access to this because it's the same as listed in sbabcock23's repo. 2. It can conflict if users have their own version of Sentry configuration in an application 3. Sentry doesn't recommend it to be initialized in a library since it can conflict with another initialized instance 4. We don't want to collect any sensitive information from the user's application.

ajacques replied:
This method itself looks like it's actually a revert from the refactoring I did. If there's a reason for the revert, happy to discuss, but right now it's just dropping those changes.

ajacques commented on custom_components/tryfi/pytryfi/__init__.py:
```diff
@@ -1,106 +1,187 @@
+                    LOGGER.debug(f"Adding Base: {b._name} Online: {b._online}")
+                    self._bases.append(b)
+                h = h + 1
+        except Exception as e:
+            capture_exception(e)
```
In a library, this style of error handling is not great because it means callers don't actually know when errors happen. We should let this bubble up and only handle errors that can reasonably be handled in a library.

ajacques commented on custom_components/tryfi/pytryfi/__init__.py:
```diff
@@ -1,106 +1,187 @@
+            for house in petListJSON:
+                for pet in house['household']['pets']:
+                    p = FiPet(pet['id'])
+                    p.setPetDetailsJSON(pet)
+                    #get the current location and set it
+                    pLocJSON = query.getCurrentPetLocation(self._session,p._petId)
+                    p.setCurrentLocation(pLocJSON)
+                    #get the daily, weekly and monthly stats and set
+                    pStatsJSON = query.getCurrentPetStats(self._session,p._petId)
+                    p.setStats(pStatsJSON['dailyStat'],pStatsJSON['weeklyStat'],pStatsJSON['monthlyStat'])
+                    #get the daily, weekly and monthly rest stats and set
+                    pRestStatsJSON = query.getCurrentPetRestStats(self._session,p._petId)
+                    p.setRestStats(pRestStatsJSON['dailyStat'],pRestStatsJSON['weeklyStat'],pRestStatsJSON['monthlyStat'])
+                    #get behavior stats if available (Series 3+ only)
```
This removes the GraphQL call batching. In the previous version, I made a single GraphQL call to get all the data necessary for the pet. Is that intentional?

ajacques commented on custom_components/tryfi/pytryfi/fiPet.py:
```diff
@@ -1,136 +1,494 @@
+            LOGGER.info(f"Pet {self._name} has no device assigned")
+
+        # Initialize behavior metrics to 0 (will be updated later if available)
+        self._dailyBarkingCount = 0
+        self._dailyBarkingDuration = 0
+        self._weeklyBarkingCount = 0
+        self._weeklyBarkingDuration = 0
+        self._monthlyBarkingCount = 0
+        self._monthlyBarkingDuration = 0
```
Optional, but one thing I was considering is just have a StatsObject and the fiPet has just a "_dailyStats", "_monthlyStats", "_weeklyStats", so we don't have to duplicate this code.

ajacques commented:
I still see a bunch of reverts, especially around error handling and the updatePets reverts the GraphQL Batching logic

ajacques commented:
Hey @adamjacobmuller That commit helped, but just skimming through the diff there's still numerous reverts in many different files. Take fiPet for example, things like me removing using exceptions as control flow which makes the code more brittle and harder to follow, removing type hints, the really interesting part is `FiPet.updateAllDetails`. 

Before:
```python
    def updateAllDetails(self, sessionId: requests.Session):
        petJson = query.getPetAllInfo(sessionId, self.petId)
        self.device.setDeviceDetailsJSON(petJson['device'])
        self.setCurrentLocation(petJson['ongoingActivity'])
        self.setStats(petJson['dailyStepStat'], petJson['weeklyStepStat'], petJson['monthlyStepStat'])
        # TODO: Support weekly
        self._dailySleep, self._dailyNap = self._extractSleep(petJson['dailySleepStat'])
        self._monthlySleep, self._monthlyNap = self._extractSleep(petJson['monthlySleepStat'])
```

After:
```python
 def updateAllDetails(self, sessionId):
        self.updateDeviceDetails(sessionId)
        self.updatePetLocation(sessionId)
        self.updateStats(sessionId)
        self.updateRestStats(sessionId)
        self.updateBehaviorStats(sessionId)
```

That undoes the GraphQL batching changes to make a single call and makes it so that clients make multiple round-trips to tryfi.com. This helps reduce the probability of tryfi wanting to block, improves performance, and helps avoid intermittent network failures. 

Just skimming through other files shows similar random reverts. I suspect we had a bit of a struggle because we both did merge into two diverging code bases initially and we both ended up having to merge each others changes in and once we get past this, then it'll be smooth sailing.

I won't claim that my code is perfect, but given the diff, it seems more like a merge mistake than intentional. One option would be to just reset to the master branch and then cherry-pick just your new changes into the PR. I do that a lot when my branches unmanageably diverged.

adamjacobmuller commented:
Hi @ajacques - you were absolutely right about the reverts\! 

I've completely redone the PR following your suggestion:
1. Started fresh from master branch
2. Added ONLY the new behavior tracking functionality
3. Preserved ALL your optimizations

The new implementation (commit 78a3384):
✅ **updateAllDetails()** still uses your batched `getPetAllInfo()` - just adds a call to fetch behavior data
✅ **updatePets()** is untouched - still uses your efficient per-pet updates  
✅ All error handling improvements preserved
✅ All type hints preserved
✅ No reverts whatsoever - purely additive changes

The behavior tracking is now implemented as:
- Added `getPetHealthTrends()` function to query.py
- Added `updateBehaviorStats()` method that's called from updateAllDetails() 
- Added behavior properties to FiPet
- Added PetBehaviorSensor class to sensor.py

All tests pass and CI should be green. Thanks for your patience and the thorough review\!

adamjacobmuller commented:
UPDATE: I've also addressed your inline review comments about Sentry (commit 249fdb0):

✅ Removed all `sentry_sdk` imports
✅ Removed all `capture_exception()` calls  
✅ Errors now bubble up naturally for proper handling

You're absolutely right that libraries shouldn't initialize Sentry - it can conflict with user applications and potentially collect sensitive data.

The PR now:
1. Adds only the new behavior tracking features
2. Preserves all your GraphQL batching and error handling improvements
3. Removes all Sentry usage
4. All tests pass

Thanks for the thorough code review\!

ajacques commented on custom_components/tryfi/pytryfi/fiPet.py:
```diff
@@ -160,6 +190,93 @@ def updateAllDetails(self, sessionId: requests.Session):
+            return True
+        except Exception as e:
+            LOGGER.debug(f"Could not update behavior stats for Pet {self.name}. This may be an older collar model.\n{e}")
+            return False
```
The updateAllInfo also has a try catch, making this handle redundant. I think it would be better to remove this, let it throw, then catch and log in updateAllInfo

ajacques commented on custom_components/tryfi/pytryfi/common/query.py:
```diff
@@ -104,6 +104,45 @@ def getDevicedetails(session: requests.Session, petId: str):
+                    eventsSummary
+                    durationSummary
+                }}
+                chart {{
```
Are we using the historical chart data in the periodic refresh? If not, then I'd suggest removing it to reduce the size of the GraphQL since it looks like a time series. Then to fetch the history, add another GraphQL call just for that. This helps keep our API response payloads small and performant.

ajacques commented on custom_components/tryfi/pytryfi/fiPet.py:
```diff
@@ -160,6 +190,93 @@ def updateAllDetails(self, sessionId: requests.Session):
         self._monthlySleep, self._monthlyNap = self._extractSleep(petJson['monthlySleepStat'])
+        # Try to fetch behavior data for Series 3+ collars
+        try:
+            self.updateBehaviorStats(sessionId)
```
Instead of making the service call, then seeing if it has any data, are we able to look at the FiDevice object and check to see if it's a series 3? That would avoid the extra call for series 1 and 2 collars.

ajacques commented on custom_components/tryfi/pytryfi/fiPet.py:
```diff
@@ -160,6 +190,93 @@ def updateAllDetails(self, sessionId: requests.Session):
+                duration_seconds = 0
+                if duration_summary:
+                    try:
+                        if duration_summary.startswith('<'):
+                            duration_seconds = 30
+                        else:
+                            parts = duration_summary.replace('h', '').replace('m', '').split()
+                            if len(parts) == 2:
+                                duration_seconds = int(parts[0]) * 3600 + int(parts[1]) * 60
```
I don't have a sample response, so I'm not quite sure what the durationSummary looks like which makes this block of code kind of confusing. Is it like 1h3m or a different format?

ajacques reviewed (commented):
Sorry I took some time to get back to this. I've been traveling for work with limited Internet access. Do you have a sample copy of the API call responses? I don't have a series 3 collar to see them. Can be anonymized if you want. Here are a few thoughts I had when reading over, feel free to consider them if you agree.

ajacques commented:
I made a few small tweaks. I hope you don't mind. I've also tested it on my series 2 and it no-ops the call, so no regression on API calls for those collars. LGTM
